app/Foliage/CmdBuild.hs

```diff
@@ -290,8 +297,48 @@ prepareIndexPkgMetadata expiryTime PreparedPackageVersion {pkgId, sdistPath} = d
         targetsDelegations = Nothing
       }
 
+prepareIndexPkgExtraEntries :: UTCTime -> [PreparedPackageVersion] -> [Tar.Entry]
+prepareIndexPkgExtraEntries timestamp packageVersions = map buildTarEntry $ Map.toList $ getPreferredVersions packageVersions
+  where
+    buildTarEntry (pkgName, pkgVersions) = mkTarEntry' content path timestamp
+      --                                                            ^^^^^^^^^ TODO: this is for now set to `currentTime` ...
+      --                                                            ... should it be the timestamp of the latest version of the package?
+      where
+        path = "index" </> prettyShow pkgName </> "preferred-versions"
+        content = generatePreferredVersionsFileContent $ map snd pkgVersions
+
+getPreferredVersions :: [PreparedPackageVersion] -> Map.Map PackageName [(Maybe UTCTime, Version)]
+--                                                                        ^^^^^^^^^^^^^ TODO: actually I don't use this value yet, should I?
+getPreferredVersions packageVersions = Map.fromListWith (++) [(getKey p, getValue p) | p <- packageVersions]
+  --                                                     ^^ TODO: replace it by `consolidateRanges`? To work with a list of `VersionRange` instead of a list of `Version`?
+  where
+    getKey PreparedPackageVersion {pkgId = PackageIdentifier {pkgName}} = pkgName
+    getValue PreparedPackageVersion {pkgTimestamp, pkgVersionDeprecated, pkgId = PackageIdentifier {pkgVersion}} = [(pkgTimestamp, pkgVersion) | not pkgVersionDeprecated]
```

Comment:
With the changes suggested in `PreparePackageVersion.hs` this should start making more sense. Each `PreparedPackageVersion` can give rise to multiple entries (one for each deprecation change).

Now we have something like `[(PackageName, [(UTCTime, Bool)]]` which we need to turn into `preferred-versions` entries which are ~ `[(UTCTime, PackageName, VersionRange)]`, where the version range has to be computed by accumulating the deprecations/undeprecations.

Spitting out a list of non deprecatated versions should work just fine I think. (i.e. treating `VersionRange` more like `[Version]`).